### PR TITLE
[Drools-3838] Improve error reporting in scenario result

### PR DESCRIPTION
**Pull request by yesamer**
@kkufova @gitgabrio @danielezonca @jomarko can you please test it and review it? Thanks!

https://issues.jboss.org/browse/DROOLS-3838

This PR contains the client sidechanges required for DROOLS-3838.
Related PR on drools-wb which contains backend change is here kiegroup/drools#2335

PRs list:
    kiegroup/drools#2335

### CODE REVIEW

**Comment by kkufova** [comment]
@yesamer, I noticed the following things:

1. First of all, I thought there should be a small red icon displayed directly in the cell for displaying the message. These plans have changed?

2. Secondly, I discovered a serious bug: if the expected value is `null` (and the test fails because the entered value is different), the following unexpected error is thrown every time you hover over the cell:
```
Uncaught exception: Exception caught: Exception caught: (TypeError) : Cannot read property 'toString' of undefined Caused by: Exception caught: (TypeError) : Cannot read property 'toString' of undefined Caused by: (TypeError) : Cannot read property 'toString' of undefined
```

3. The same things happens with imported test scenarios. If you import a {{.scesim}} file, instead of displaying the message, an unexpected error is thrown when there's an error.

4. I noticed that in the window, the **Apply** button is dark grey as opposed to [the original design](https://issues.jboss.org/secure/attachment/12447628/Error%20reporting%20after%20test%20run-popup.png). Can this be fixed?

**Comment by yesamer** [comment]
@kkufova Points 2,3,4 can be easily managed. Point 1 could be a little bit harder to implement.

**Comment by danielezonca** [comment]
@kkufova @yesamer Please manage point 1 with a different ticket

**Comment by yesamer** [comment]
I created https://issues.jboss.org/browse/DROOLS-3989 to manage point 1.
Point 2,3 and 4 are fixed.

**Comment by kkufova** [comment]
@yesamer, I found the following issues:

1. If a cell is highlighted, you fix the value and rerun the tests, the cell is still highlighted even though the tests are now passing. Can this be fixed?

2. I noticed that `Impossible to parse as boolean` message keeps getting displayed instead of an actual help. For example, I used `falsee` instead of `false`. I think in this case we definitely need `The expected value is 'false' but the actual one is 'falsee'.` Another scenario: if the expected value is `null` but I enter `[]` instead, the error message is `Error with this value: Impossible to parse as boolean`. It doesn't say anything about the actual value being `null`.

3. If the expected value is `null` and you use something else (such as `true`) and then click **Apply**, instead of adding `null`, the cell is empty.

All three bugs can be seed [here](https://issues.jboss.org/secure/attachment/12450417/screencast-07-05-19-1.webm).

**Comment by yesamer** [comment]
@jomarko Hi Jozef, I fixed points 1 and 3 raised by @kkufova. Point 2 should be managed in another ticket in my opinion. Please consider that the changes contained in https://github.com/kiegroup/drools/pull/2335 are mandatory to test it.
Thanks.

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+public class CommonOnMoveHandler extends AbstractScenarioSimulationGridPanelHandler {
+
+    /* This parameter must be synchronized with POPOVER_WIDTH static variable in ErrorReportPopoverView.less */
+    private static int POPOVER_WIDTH = 200;
+    private static String NULL = "null";
+
+    protected ErrorReportPopoverPresenter errorReportPopupPresenter;
+
+    protected Integer currentlyShownHeaderRowIndex = -1;
+    protected Integer currentlyShownHeaderColumnIndex = -1;
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java]: `currentlyShownHeaderRowIndex` and `currentlyShownHeaderColumnIndex` are never used, is it intentional?

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/AbstractScenarioSimulationGridPanelHandler.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        if (uiRowIndex == null) {
+            uiRowIndex = getUiRowIndexLocal(gridClickPoint.getY());
+            isHeader = false;
+        }
+        final Integer uiColumnIndex = getUiColumnIndexLocal(gridClickPoint.getX());
+        ScenarioGridColumn scenarioGridColumn = uiColumnIndex != null ? (ScenarioGridColumn) scenarioGrid.getModel().getColumns().get(uiColumnIndex) : null;
+        if (isHeader) {
+            return manageHeaderCoordinates(uiColumnIndex, scenarioGridColumn, gridClickPoint);
+        } else {
+            return (uiRowIndex == null || uiColumnIndex == null) ? manageBodyCoordinates(-1, -1) : manageBodyCoordinates(uiRowIndex, uiColumnIndex);
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/AbstractScenarioSimulationGridPanelHandler.java]: I would change this line to simple `return manageBodyCoordinates(uiRowIndex, uiColumnIndex)` and add the check for null values inside of `manageBodyCoordinates`. This wil facilitate invocation from other places (if needed in future), what do you think?

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+    @Override
+    protected boolean manageGivenExpectHeaderCoordinates(ScenarioHeaderMetaData clickedScenarioHeaderMetadata, ScenarioGridColumn scenarioGridColumn, String group, Integer uiColumnIndex) {
+        return false;
+    }
+
+    @Override
+    protected boolean manageBodyCoordinates(Integer uiRowIndex, Integer uiColumnIndex) {
+        /* In this case, the mouse is out ot the GridLayer, then return false, without perform any action */
+        if (uiColumnIndex == -1 || uiRowIndex == -1) {
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java]: Related to comment in `AbstractScenarioSimulationGridPanelHandler`. This `if` would change to `uiColumnIndex == null || uiRowIndex == null`

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        final Optional<FactMappingValue> factMappingValueOptional = scenarioByIndex.getFactMappingValue(factMapping);
+        factMappingValueOptional.ifPresent(factMappingValue -> {
+            /* If an error is present in the FactMappingValue, it calculates the coordinates for Popover and show it */
+            if (FactMappingValueStatus.SUCCESS != factMappingValue.getStatus()) {
+                /* It updates the coordinates of the current shown cell */
+                currentlyShownBodyRowIndex = uiRowIndex;
+                currentlyShownBodyColumnIndex = uiColumnIndex;
+                /* It calculates the coordinates */
+                final GridColumn<?> column = scenarioGrid.getModel().getColumns().get(uiColumnIndex);
+                Point2D xYCell = retrieveCellMiddleXYPosition(column, uiRowIndex);
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java]: `xYCell` can be also final

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                                                                                         uiColumnIndex,
+                                                                                         errorValue != null ? errorValue.toString() : NULL));
+                                                   },
+                                                   xPosition,
+                                                   yPosition,
+                                                   position);
+                } else {
+                    errorReportPopupPresenter.show(ScenarioSimulationEditorConstants.INSTANCE.errorReason(),
+                                                   ScenarioSimulationEditorConstants.INSTANCE.errorPopoverMessageFailedWithException(
+                                                           factMappingValue.getExceptionMessage()),
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java]: Do we need to check `null` value here as we do for `FAILED_WITH_ERROR` and if test is positive use constant `NULL`?

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/ScenarioInvokeContextMenuForSelectedCell.java [hunk 6]
```diff
@@ -68,30 +68,10 @@ public boolean perform(final GridWidget gridWidget, final boolean isShiftKeyDown
...
-                                                                 ci,
-                                                                 uiRowIndex);
-            final int cellXMiddle = (int) (context.getAbsoluteCellX() +
-                    context.getCellWidth() / 2 +
-                    gridLayer.getDomElementContainer().getAbsoluteLeft());
-            final int cellYMiddle = (int) (context.getAbsoluteCellY() +
-                    context.getCellHeight() / 2 +
-                    gridLayer.getDomElementContainer().getAbsoluteTop());
-
+            final Point2D middleXYCell = getMiddleXYCell(gridWidget, column, isHeader, uiRowIndex, gridLayer);
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/ScenarioInvokeContextMenuForSelectedCell.java]: Happy to see code reuse, perfect!

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/menu/ScenarioContextMenuRegistry.java [hunk 7]
```diff
@@ -255,4 +257,14 @@ private boolean manageHeaderRightClick(final ScenarioGrid scenarioGrid,
         scenarioGrid.setSelectedColumnAndHeader(uiHeaderRowIndex, uiColumnIndex);
         return true;
     }
+
+
+    public void setErrorReportPopupPresenter(ErrorReportPopoverPresenter errorReportPopupPresenter) {
+        this.errorReportPopupPresenter = errorReportPopupPresenter;
+    }
+
+    public void hideErrorReportPopupover() {
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/menu/ScenarioContextMenuRegistry.java]: Just double checking naming, is `hideErrorReportPopupover` correct, the `over` suffix seems to be not needed to me.

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/popover/AbstractPopoverView.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+                               final Div popoverContentElement,
+                               final JQueryProducer.JQuery<Popover> jQueryPopover) {
+        this.popoverElement = popoverElement;
+        this.popoverContainerElement = popoverContainerElement;
+        this.popoverContentElement = popoverContentElement;
+        this.jQueryPopover = jQueryPopover;
+    }
+
+    @Override
+    public void show(final Optional<String> editorTitle, final int mx, final int my, Position position) {
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/popover/AbstractPopoverView.java]: Just for consistency, please make also the last parameter final: `final Position position`

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/popover/ErrorReportPopoverView.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        super.show(Optional.of(errorTitleText), mx, my, position);
+    }
+
+    @Override
+    public void show(final String errorTitleText,
+                         final String errorContentText,
+                         final String keepText,
+                         final int mx,
+                         final int my,
+                         final Position position) {
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/popover/ErrorReportPopoverView.java]: Just formatting - indent of parameters.

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties [hunk 10]
```diff
@@ -146,5 +146,10 @@ reportAvailableLabel=Available decisions:
...
 numberOfTimesDecisionEvaluated=A number of times each decision has been evaluated:
-
-
+#
+errorReason=Error reason
+keep=Keep
+close=Close
+apply=Apply
+errorPopoverMessageFailedWithError=The expected value is ''{0}'' but the actual one is ''{1}''.
+errorPopoverMessageFailedWithException=Error with this value: {0}
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties]: Not sure if this comes form UX mockups, but I would expect something more like -  "Exception with detailed message: {0}"

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/test/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandlerTest.java [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+    private CommonOnMoveHandler commonOnMoveHandler;
+
+    @Mock
+    private ErrorReportPopoverPresenter errorReportPopupPresenterMock;
+    @Mock
+    private Scenario scenarioMock;
+    @Mock
+    private FactMappingValue factMappingValueMock;
+    @Mock
+    private AbsolutePanel scrollPanel;
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/test/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandlerTest.java]: to be consistent, please use name `scrollPanelMock`

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/test/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandlerTest.java [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+    @Mock
+    private Element element;
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/test/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandlerTest.java]: To be consistent, please use name `elementMock`

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties (hunk 10 above)
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties]: In my knowledge, this case is not handled in UX mockups.
Fine for me to put yours.
DONE!

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/test/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandlerTest.java (hunk 11 above)
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/test/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandlerTest.java]: Done!

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/test/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandlerTest.java (hunk 12 above)
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/test/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandlerTest.java]: Done!

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/popover/ErrorReportPopoverView.java (hunk 9 above)
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/popover/ErrorReportPopoverView.java]: Done!

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/popover/AbstractPopoverView.java (hunk 8 above)
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/popover/AbstractPopoverView.java]: DONE!

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/menu/ScenarioContextMenuRegistry.java (hunk 7 above)
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/menu/ScenarioContextMenuRegistry.java]: The correct name is hideErrorReportPopover.
DONE

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java (hunk 4 above)
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java]: Done!

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java (hunk 1 above)
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java]: At the moment these are not used, you right. Could be used in the future when a popover for the header cells is required. But I agree with you, they can be removed. 
Done!

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/AbstractScenarioSimulationGridPanelHandler.java (hunk 2 above)
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/AbstractScenarioSimulationGridPanelHandler.java]: Agree, DONE!

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java (hunk 3 above)
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java]: Agree, DONE!

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java (hunk 5 above)
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/CommonOnMoveHandler.java]: Do you mean to test if <factMappingValue.getStatus()> is null?
Yes, I agree to add it, but in another point. In my opinion, if that value is NULL, the popover shouldn't be visible at all. I put it in line 87.

**Comment by jomarko** [comment]
jenkins execute full downstream build

**Comment by jomarko** [comment]
@yesamer thank you for the PR, works very nice. I can confirm pints 1. and 3. from [comment](https://github.com/kiegroup/drools-wb/pull/1125#issuecomment-490096342) are fixed. The point 2. will be addressed by [DROOLS-3992](https://issues.jboss.org/browse/DROOLS-3992). There remain two points that I would like to discuss:

1. The text we show to user when some exception occurred during parsing cell value. The popover title is definitely fine - **Error reason**. However I think we should still improve the prefix of the exception message to make it maybe shorter and more clear? I think it is question to UX of your scrum team? What do you think? [1]
2. If some issue was in multiple ceels of one row, and user decides to fix one value (**Apply** button is pressed) whole row highlight is lost. See [2]. Is this expected or not? if not, should I file a separate jira or worth of adding into this PR?

[1]
Screenshot from 2019-05-10 09-18-21

[2]
https://drive.google.com/file/d/1c-Hgc6ePD2OttI-ddVLhqj9UU0psrLPF/view?usp=sharing

**Comment by yesamer** [comment]
@jomarko Thank you!
About your considerations:
1) It could be a solution to involve UX team to find a better message. We can handle this in DROOLS-3992, in my opinon.
2) I didn't managed the case you reported, then I agree with you to open a new ticket. Please assign it to me, I will manage it in the next sprint.
Thank you for the review and the test

**Comment by danielezonca** [comment]
jenkins execute full downstream build

**Comment by yesamer** [comment]
Hi @manstis, can you please merge this PR and kiegroup/drools#2335?
Thanks,

**Pull request merged**